book-content/chapters/06-objects.md

```diff
@@ -0,0 +1,1386 @@
+So far, we've looked at object types only in the context of 'object literals', defined using `{}` with type aliases.
+
+But TypeScript has many tools available that let you be more expressive with object types. You can model inheritance, create new object types from existing ones, and use dynamic keys.
+
+## Extending Objects
+
+Let's start our investigation by looking at how to build object types from _other object types_ in TypeScript.
+
+### Intersection Types
+
+An intersection type lets us combine multiple object types into a single type. It uses the `&` operator. You can think of it like the reverse of the `|` operator. Instead of representing an "or" relationship between types, the `&` operator signifies an "and" relationship.
+
+Using the intersection operator `&` combines multiple separate types into a single type.
+
+Consider these types for `Album` and `SalesData`:
+
+```typescript
+type Album = {
+  title: string;
+  artist: string;
+  releaseYear: number;
+};
+
+type SalesData = {
+  unitsSold: number;
+  revenue: number;
+};
+```
+
+On their own, each type represents a distinct set of properties. While the `SalesData` type on its own could be used to represent sales data for any product, using the `&` operator to create an intersection type allows us to combine the two types into a single type that represents an album's sales data:
+
+```typescript
+type AlbumSales = Album & SalesData;
+```
+
+The `AlbumSales` type now requires objects to include all of the properties from both `AlbumDetails` and `SalesData`:
+
+```typescript
+const wishYouWereHereSales: AlbumSales = {
+  title: "Wish You Were Here",
+  artist: "Pink Floyd",
+  releaseYear: 1975
+  unitsSold: 13000000,
+  revenue: 65000000,
+};
+```
+
+If the contract of the `AlbumSales` type isn't fulfilled when creating a new object, TypeScript will raise an error.
+
+It's also possible to intersect more than two types:
+
+```typescript
+type AlbumSales = Album & SalesData & { genre: string };
+```
+
+This is a useful method for creating new types from existing ones.
+
+#### Intersection Types With Primitives
+
+It's worth noting that intersection types can also be used with primitives, like `string` and `number` - though it often produces odd results.
+
+For instance, let's try intersecting `string` and `number`:
+
+```typescript
+type StringAndNumber = string & number;
+```
+
+What type do you think `StringAndNumber` is? It's actually `never`. This is because `string` and `number` have innate properties that can't be combined together.
+
+This also happens when you intersect two object types with an incompatible property:
+
+```ts twoslash
+type User1 = {
+  age: number;
+};
+
+type User2 = {
+  age: string;
+};
+
+type User = User1 & User2;
+//   ^?
+```
+
+In this case, the `age` property resolves to `never` because it's impossible for a single property to be both a `number` and a `string`.
```

Comment:
but it's only the `age` property that resolves to `never` - the type itself, its potential other properties, etc wont be never-ified. That makes the previous definition that you coined above inaccurate:
> What type do you think `StringAndNumber` is? It's actually `never`. This is because `string` and `number` have innate properties that can't be combined together.

In that case it's not about mismatched properties but about completely disjoint domains (or something ;p). Like, there is maybe some overlap between both of those that could be the output of such operation. For example, both `number` and `string` have `{ toString(): string }`. For practical purposes, it's rather useless to create an object with only compatible methods though.